src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1962,6 +1963,151 @@ static TypeDeclarationSyntax constructTypeDeclaration(ContextAwareSyntax syntaxF
             }
         }
 
+        private MemberDeclarationSyntax ParseExtensionContainer(SyntaxList<AttributeListSyntax> attributes, SyntaxListBuilder modifiers)
+        {
+            Debug.Assert(!IsInAsync);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.ExtensionKeyword);
+            SyntaxToken extensionKeyword = this.EatContextualToken(SyntaxKind.ExtensionKeyword);
+
+            var outerSaveTerm = _termState;
+            _termState |= TerminatorState.IsPossibleAggregateClauseStartOrStop; // TODO2 add test affected
+
+            TypeParameterListSyntax typeParameters = this.ParseTypeParameterList();
+            ReceiverParameterSyntax receiverParameter = parseReceiverParameter(out var openParenToken, out var closeParenToken);
+
+            // Parse extension container body
+            bool parseMembers = true;
+            SyntaxListBuilder<TypeParameterConstraintClauseSyntax> constraints = default;
+            SyntaxListBuilder<MemberDeclarationSyntax> members = default;
+            try
+            {
+                if (this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword)
+                {
+                    _termState |= TerminatorState.IsEndOfRecordOrClassOrStructOrInterfaceSignature;
+                    constraints = _pool.Allocate<TypeParameterConstraintClauseSyntax>();
+                    this.ParseTypeParameterConstraintClauses(constraints);
+                }
```

Comment:
extract this into helper.  otehrwise it can get out of sync with type parameter parsing elsewhere.